x/leverage/keeper/msg_server_test.go

```diff
@@ -172,3 +173,1545 @@ func (s *IntegrationTestSuite) TestUpdateRegistry() {
+
+	// create a DUMP (historic price 1.00, current price 0.50) borrower
+	// using PUMP (historic price 1.00, current price 2.00) collateral
+	dumpborrower := s.newAccount(coin(pumpDenom, 100_000000))
+	s.supply(dumpborrower, coin(pumpDenom, 100_000000))
+	s.collateralize(dumpborrower, coin("u/"+pumpDenom, 100_000000))
+	s.borrow(dumpborrower, coin(dumpDenom, 20_000000))
+	// collateral value is $200 (current) or $100 (historic)
+	// borrowed value is $10 (current) or $20 (historic)
+	// collateral weights are always 0.25 in testing
+
+	// create a PUMP (historic price 1.00, current price 2.00) borrower
+	// using DUMP (historic price 1.00, current price 0.50) collateral
+	pumpborrower := s.newAccount(coin(dumpDenom, 100_000000))
+	s.supply(pumpborrower, coin(dumpDenom, 100_000000))
+	s.collateralize(pumpborrower, coin("u/"+dumpDenom, 100_000000))
+	s.borrow(pumpborrower, coin(pumpDenom, 5_000000))
+	// collateral value is $50 (current) or $100 (historic)
+	// borrowed value is $10 (current) or $5 (historic)
+	// collateral weights are always 0.25 in testing
+
+	tcs := []testCase{
+		{
+			"unregistered base token",
+			supplier,
+			"abcd",
+			sdk.Coin{},
+			sdk.Coin{},
+			sdk.Coin{},
+			types.ErrMaxWithdrawZero,
+		},
+		{
+			"uToken",
+			supplier,
+			"u/" + umeeDenom,
+			sdk.Coin{},
+			sdk.Coin{},
+			sdk.Coin{},
+			types.ErrUToken,
+		},
+		{
+			"max withdraw umee",
+			supplier,
+			umeeDenom,
+			coin("u/"+umeeDenom, 100_000000),
+			coin("u/"+umeeDenom, 75_000000),
+			coin(umeeDenom, 100_000000),
+			nil,
+		},
+		{
+			"duplicate max withdraw umee",
+			supplier,
+			umeeDenom,
+			sdk.Coin{},
+			sdk.Coin{},
+			sdk.Coin{},
+			types.ErrMaxWithdrawZero,
+		},
+		{
+			"max withdraw with borrow",
+			other,
+			umeeDenom,
+			coin("u/"+umeeDenom, 60_000000),
+			coin("u/"+umeeDenom, 60_000000),
+			coin(umeeDenom, 60_000000),
+			nil,
+		},
+		{
+			"max withdrawal (dump borrower)",
+			dumpborrower,
+			pumpDenom,
+			coin("u/"+pumpDenom, 20_000000),
+			coin("u/"+pumpDenom, 20_000000),
+			coin(pumpDenom, 20_000000),
+			nil,
+		},
+		{
+			"max withdrawal (pump borrower)",
+			pumpborrower,
+			dumpDenom,
+			coin("u/"+dumpDenom, 20_000000),
+			coin("u/"+dumpDenom, 20_000000),
+			coin(dumpDenom, 20_000000),
+			nil,
+		},
+	}
+
+	for _, tc := range tcs {
+		msg := &types.MsgMaxWithdraw{
+			Supplier: tc.addr.String(),
+			Denom:    tc.denom,
+		}
+		if tc.err != nil {
+			_, err := srv.MaxWithdraw(ctx, msg)
+			require.ErrorIs(err, tc.err, tc.msg)
+		} else {
+			expectFromBalance := tc.expectedWithdraw.Sub(tc.expectFromCollateral)
+
+			// initial state
+			iBalance := app.BankKeeper.GetAllBalances(ctx, tc.addr)
+			iCollateral := app.LeverageKeeper.GetBorrowerCollateral(ctx, tc.addr)
+			iUTokenSupply := app.LeverageKeeper.GetAllUTokenSupply(ctx)
+			iExchangeRate := app.LeverageKeeper.DeriveExchangeRate(ctx, tc.denom)
+			iBorrowed := app.LeverageKeeper.GetBorrowerBorrows(ctx, tc.addr)
```

Comment:
maybe we can have a helper function:
```
type balances struct {
  balance, collateral, utokenSupply, exchangeRate, borrowed
}
func getBalances(ctx, addr) balances
```
And use it across test functions : 
```
before := getBalances(...)
...
after := getBalances(...)
```
Will make the code easier to read (and much shorter)